lms/djangoapps/courseware/date_summary.py

```diff
@@ -240,13 +256,45 @@ def is_enabled(self):
 
     @lazy
     def date(self):
+        deadline = None
+
         try:
-            verified_mode = CourseMode.objects.get(
-                course_id=self.course_id, mode_slug=CourseMode.VERIFIED
-            )
-            return verified_mode.expiration_datetime
+            verified_mode = CourseMode.objects.get(course_id=self.course_id, mode_slug=CourseMode.VERIFIED)
+            deadline = verified_mode.expiration_datetime
         except CourseMode.DoesNotExist:
-            return None
+            pass
+
+        if self.course and self.course_overview.self_paced and self.enrollment:
+            # Check if the given course has opted out of the feature
+            course_config = CourseDynamicUpgradeDeadlineConfiguration.current(self.course.id)
+
+            # NOTE: We check the pk here to determine if we have an instance generated from the current() method or
+            # pulled from the database. If the instance isn't from the database, it is not relevant.
```

Comment:
Can you avoid this extra (implementation-specific) logic _(pulled from the db or not)_ if you check the global flag `DynamicUpgradeDeadlineConfiguration` before checking the course-specific one?  Something like:

```
global_config = DynamicUpgradeDeadlineConfiguration.current()
if global_config.enabled:
    delta = global_config.deadline_days
    course_config = CourseDynamicUpgradeDeadlineConfiguration.current(self.course.id)
    if course_config.enabled:
        if course_config.opt_out:
            return deadline
        delta = course_config.deadline_days
    # continue here
```

Also, I'd imagine we'd want the global kill-switch to disable all of our code (in case there's any logic/implementation issues).

Or, better yet, hide this logic within a `classmethod` on `CourseDynamicUpgradeDeadlineConfiguration`.  And then all this config-checking code can be abstracted within the config class/module.